cql3/statements/describe_statement.cc

```diff
@@ -537,8 +538,34 @@ future<std::vector<description>> describe_all_keyspace_elements(const data_dicti
     co_return result;
 }
 
+future<std::vector<description>> describe_auth_and_service_levels(const auth::service& auth_service,
+        const qos::service_level_controller& sl_controller, const data_dictionary::database& db)
+{
+    auto& replica_db = db.real_database();
+    const auto& role_manager = auth_service.underlying_role_manager();
+    const auto& authorizer = auth_service.underlying_authorizer();
+
+    auto role_descs = co_await generate_descriptions(replica_db, co_await role_manager.get_role_info(true), true);
+    auto grant_descs = co_await generate_descriptions_with_multiple_generators(co_await role_manager.get_role_grants(), replica_db, true);
+    auto permission_descs = co_await generate_descriptions_with_multiple_generators(co_await authorizer.list_all(), replica_db, true);
+    auto attached_sl_descs = co_await generate_descriptions(replica_db, co_await role_manager.get_attached_service_levels(), true);
+
+    auto service_level_descs = co_await generate_descriptions_with_generator(co_await sl_controller.get_distributed_service_levels(), replica_db, true);
```

Comment:
To avoid doing a query, you can use information about service levels from controller's cache (field: `_service_levels_db`). 
With raft and recent changes to SL, the cache gets updated while applying raft log, so the cache state is up-to-date with the table.

Reply:
Thank you! This is very useful information